File: 6.Software-Design-Patterns/example/Creational/Object-Pool/Object-Pool.js
class Connection {
  constructor() {
    console.log("Creating a new connection...");
  }
  connect() {
    console.log("Connected!");
  }
}
class ConnectionPool {
  constructor(maxSize) {
    this.maxSize = maxSize;
    this.pool = [];
  }
  getConnection() {
    if (this.pool.length >= this.maxSize) {
      throw new Error("Maximum poll size reached");
    }
    if (this.pool.length > 0) {
      const connection = this.pool.pop();
      console.log("Obtained connection from pool");
      return connection;
    }
    return new Connection();
  }
  releaseConnection(connection) {
    console.log("Releasing connection back to pool");
    this.pool.push(connection);
  }
}

const pool = new ConnectionPool(2);
const connection1 = pool.getConnection();
console.log("connection1:", connection1);

connection1.connect();
pool.releaseConnection(connection1);

const connection2 = pool.getConnection();
connection2.connect();
pool.releaseConnection(connection2);
const connection3 = pool.getConnection();
connection3.connect();

pool.releaseConnection(connection3);
const connection4 = pool.getConnection();
connection4.connect();
pool.releaseConnection(connection4);

File: 6.Software-Design-Patterns/example/Creational/Object-Pool/Object-Pool.py
class Connection:
    def __init__(self):
        print("Creating a new connection...")
    def connect(self):
        print("Connected!")
class ConnectionPool:
    def __init__(self,maxSize):
        self.pool = []
        self.maxSize = maxSize
    def get_connection(self):
        if len(self.pool) >= self.maxSize:
                raise ValueError("Maximum pool size reached")
        if self.pool:
            connection = self.pool.pop()
            print("Obtained connection from pool")
            return connection
        connection = Connection()
        return connection
    def release_connection(self,connection):
        print("Releasing connection back to pool")
        self.pool.append(connection)

pool = ConnectionPool(2)
connection1 = pool.get_connection()
print("connection1:", connection1)
connection1.connect()
pool.release_connection(connection1)

connection2 = pool.get_connection()
connection2.connect()
pool.release_connection(connection2)

connection3 = pool.get_connection()
connection3.connect()
pool.release_connection(connection3)

connection4 = pool.get_connection()
connection4.connect()
pool.release_connection(connection4)